massgov/mayflower-patternlab
# DP-8341: Implement visual regression testing for pages and templates

Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)

## Description
Implements visual regression testing using BackstopJS driving a headless Chrome browser for all Page and Template patterns.   [Successful comparison report](https://2975-62401908-gh.circle-artifacts.com/0/backstop/reports/html/index.html).  [Failed comparison report](https://2980-62401908-gh.circle-artifacts.com/0/backstop/reports/html/index.html)

Note: When this is merged, announce in #massgov:
> Starting now, Mayflower is getting visual regression testing.  If you work on Mayflower, please read the documentation at https://github.com/massgov/mayflower/blob/dev/docs/visual-regression.md.  All PRs that change the visual output of Page or Template components must update the reference screenshots.  Additionally, please ensure that your in-progress PRs are synced with the `dev` branch.  Contact Rob Bayliss with any questions. 

## Related Issue / Ticket

- [JIRA issue](DP-8341)

## Steps to Test


1. Create a new branch off of this one, make a CSS change that will impact one or more of the page or template patterns, and commit it.  
2. Observe a CI failure due to the visual differences in the pattern(s).  
3. Observe that you can view a report of the differences by viewing report in the artifacts of the "test" job.  

## Screenshots
Where the report is located:
![screen shot 2018-03-26 at 3 14 47 pm](https://user-images.githubusercontent.com/654407/37927564-795ded72-3108-11e8-8e9a-de240aeda538.png)



## Additional Notes:
N/A

#### Impacted Areas in Application


* No visual impacts.  Testing procedure only.

#### @TODO


* Once the circle 2.0 PR is merged, pull in the work from integration/vrt to actually make this work in CircleCI.

#### Today I learned...


* https://garris.github.io/BackstopJS/
* Also tried out Webdriver.io with the visual regression service, but it was too slow (took over 10 minutes to run through 60 patterns locally).  POC is here: https://github.com/massgov/mayflower/compare/poc_wdio.  This might be useful for an end-to-end test, but it's not a good candidate for CI.
